python/events/src/hydrotools/events/baseflow/eckhardt.py

```diff
@@ -0,0 +1,288 @@
+            Influenced by basin size, precipitation frequency, and other catchment
+            characteristics.
+            
+        Returns
+        -------
+        maximum_baseflow_index: float
+            The maximum baseflow index, $BFI_{max}$, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.25 and 0.8 for daily streamflow.
+            Influenced by aquifer porosity, streamflow frequency, and other catchment
+            characteristics.
+    """
+    # Instantiate maximum baseflow series
+    # Assume last value is baseflow
+    baseflow = np.empty(series.size)
+    baseflow[0] = series[-1]
+
+    # Apply reverse filter and compute
+    #   maximum baseflow index
+    flipped = series[::-1]
+    for i in range(1, series.size):
+        baseflow[i] = min(flipped[i], baseflow[i-1] / recession_constant)
+    return np.sum(baseflow) / np.sum(series)
+
+@jit(float64[:](float64[:], float64, float64), nogil=True)
+def apply_filter(
+        series: npt.NDArray,
+        recession_constant: float,
+        maximum_baseflow_index: float
+) -> npt.NDArray:
+    """Applies a digitial recursive baseflow separation filter to series
+        and returns the result.
+    
+        Parameters
+        ----------
+        series: array-type, required
+            A numpy array of streamflow values. Assumes first value in series is baseflow.
+        recession_constant: float, required
+            Linear reservoir recession constant, a, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.85 and 0.95 for daily streamflow.
+            Influenced by basin size, precipitation frequency, and other catchment
+            characteristics.
+        maximum_baseflow_index: float
+            The maximum baseflow index, $BFI_{max}$, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.25 and 0.8 for daily streamflow.
+            Influenced by aquifer porosity, streamflow frequency, and other catchment
+            characteristics.
+            
+        Returns
+        -------
+        baseflow: array-type
+            An array containing the separated baseflow values.
+    """
+    # Compute filter parameters
+    denominator = 1 - recession_constant * maximum_baseflow_index
+    A = (1.0 - maximum_baseflow_index) * recession_constant / denominator
+    B = (1.0 - recession_constant) * maximum_baseflow_index / denominator
+
+    # Instantiate baseflow series
+    # Assume first value is baseflow
+    baseflow = np.empty(series.size)
+    baseflow[0] = series[0]
+
+    # Apply filter and return result
+    for i in range(1, series.size):
+        baseflow[i] = min(series[i], A * baseflow[i-1] + B * series[i])
+    return baseflow
+
+def separate_baseflow(
+        series: pd.Series,
+        output_time_scale: Union[pd.Timedelta, datetime.timedelta, np.timedelta64, str, int],
```

Comment:
Nit, but you might consider using the newer union type hint. For backwards compatibility it requires that you import `from __future__ import annotations` as the first import in the file but IMO its worth it. This also removes the need to use `typing.List` or other collection types. You could instead just, `x: list[str] = ["cool!"]` for example.

 ```suggestion
        output_time_scale: pd.Timedelta | datetime.timedelta | np.timedelta64 | str | int,
```

Reply:
Updated.

Reply:
Note: This change meant the end of testing for 3.8 and 3.9.